### ISSUE TITLE
mbedtls implementation

### ISSUE THREAD

**Issue by MabezDev** [sfackler/rust-native-tls#211]
Without `rust-native-tls` support, a lot of std crates become partially, or completely unusable. For our std target, `riscv32imc-esp-espidf`, we have an mbedtls implementation available.

[`rust-mbedtls`](https://github.com/fortanix/rust-mbedtls) is a stable and well maintained crate with bindings to mbedtls. Would you be open to PR's adding support in `rust-native-tls`?

**Comment by sfackler** [comment]
How would you classify the set of targets that should use mbedtls?

**Comment by MabezDev** [comment]
For our use case, we can use `target_os = "espidf"`.

**Comment by sfackler** [comment]
And OpenSSL will not build at all for espidf?

**Comment by MabezDev** [comment]
We had partial bindings for openSSL but they were buggy and are being removed with esp-idf 5.0. There is a possibility we could extract the openSSL wrapper into its own library and maintain it separately.

**Comment by sfackler** [comment]
I'm just asking to confirm that making this change won't break people that were using OpenSSL on that OS already - if it's not supported at all then it should be fine to add mbedts for that case.

**Comment by MabezDev** [comment]
Correct, it won't break any existing code by adding mbedtls support for espidf in this crate.

I'll start working on a PR :).

**Comment by jan-br** [comment]
Any updates on this? :) @MabezDev

**Comment by MabezDev** [comment]
@jan-br I have it working locally, but require some changes to mbedtls. Similar patches are in review upstream in mbedtls so I'm waiting for them to get merged, before I can open a PR here :).

**Comment by jan-br** [comment]
Thats great, thanks for your work. :)
This could be quite a game changer for rust development with esp-idf.
Also just wondering, is this working with Xtensa too or is it just RiscV?

**Comment by MabezDev** [comment]
Both Xtensa and RISC-V will be supported :).

**Comment by andresv** [comment]
Almost excellent timing. I also would like to use it for `mqttrust`.

**Comment by MabezDev** [comment]
Sorry not everything is pushed at the moment. The idea is to replace mbedtls-sys-auto completely with esp-idf-sys, but its been a bit tricky to work out all the issues thus far. Once I have something I can build I'll report back here.

If you are desperate for tls however, just this weekend someone got the ring crate building on the espidf platform: https://github.com/briansmith/ring/pull/1459, with an demo here: https://github.com/killyourphone/tlsdemo.

There are still advantages to having mbedtls support (and rust-native-tls support) such as better use of the hardware peripherals for accelerated hashing etc, so I won't stop working on this but there is an alternative if you need something now :)

**Comment by jan-br** [comment]
> > Yeah TLS is an absolut hard requirement for my current project. So its been quite a blocker for a while now. I'll check the ring PR out. I could not ask for anything more if that works. That would solve alot of problems for me right now. Thanks alot for the hint!
> 
> Just saw this issue referenced from that PR after going to go check on it. Just be aware you'll have to deal with a lot of patched crates if you want to go with my ring/rustls setup, but it does work if you need something now. I also will make no guarantees of safety, stability, etc. I closely guard my ssh keys but I still never recommend people trust a stranger not to sneak something into one of your dependencies. I'm really hoping some of these changes get upstreamed so the list of crates to patch gets shorter. And of course as soon as this mbedtls work is done I think that'll be ideal for most use-cases.

Yes I saw that. It is actually quite a mess, but I think I got ring working for now.

**Comment by Janrupf** [comment]
@MabezDev are you still working on this? I recently got `mio` and `tokio` running the ESP and now I'm looking for a TLS implementation. I suppose for now one could hack something together with either `ring` or the ESP TLS API, but a proper wrapper would be amazing!

**Comment by lu-zero** [comment]
@MabezDev another poke ^^

**Comment by MabezDev** [comment]
Sorry for missing these! I'm not currently working mbedtls support but I did talk to a few colleagues internally about this. Instead of mbedtls we're probably going to wrap the esp-tls layer instead as the API surface is far smaller and in theory, allows us to use another TLS implementation underneath (WolfSSL is supported by esp-tls). I don't have any time scale on this sadly, this is currently a side project for the folks already maintaining esp-tls. For now, the best option is still to use the patched ring from [esp-rs-compat](https://github.com/esp-rs-compat/ring).

**Comment by ivmarkov** [comment]
> @MabezDev are you still working on this? I recently got `mio` and `tokio` running the ESP and now I'm looking for a TLS implementation. I suppose for now one could hack something together with either `ring` or the ESP TLS API, but a proper wrapper would be amazing!

I would be very very interested in seeing how you implemented support for `mio` on top of ESP IDF given that `mio` (a) does not support neither `select` nor `poll` and (b) relies on edge-triggering rather than level triggering. This is by the way the reason why I looked into [supporting]( https://github.com/esp-rs-compat/polling) the `polling` crate from the `smol-rs` ecosystem instead.

**Comment by Janrupf** [comment]
> > @MabezDev are you still working on this? I recently got `mio` and `tokio` running the ESP and now I'm looking for a TLS implementation. I suppose for now one could hack something together with either `ring` or the ESP TLS API, but a proper wrapper would be amazing!
> 
> @Janrupf I would be very very interested in seeing how you implemented support for `mio` on top of ESP IDF given that `mio` (a) does not support neither `select` nor `poll` and (b) relies on edge-triggering rather than level triggering. This is by the way the reason why I looked into [supporting](https://github.com/esp-rs-compat/polling) the `polling` crate from the `smol-rs` ecosystem instead.

https://github.com/tokio-rs/mio/pull/1602 - see here. TL;DR: I'm still working on a fully proper implementation (though CI is the main problem right now), but I already have a fully working mio for ESP32 implementation (and thus Tokio for ESP32)

**Comment by ivmarkov** [comment]
> > > @MabezDev are you still working on this? I recently got `mio` and `tokio` running the ESP and now I'm looking for a TLS implementation. I suppose for now one could hack something together with either `ring` or the ESP TLS API, but a proper wrapper would be amazing!
> > 
> > 
> > @Janrupf I would be very very interested in seeing how you implemented support for `mio` on top of ESP IDF given that `mio` (a) does not support neither `select` nor `poll` and (b) relies on edge-triggering rather than level triggering. This is by the way the reason why I looked into [supporting](https://github.com/esp-rs-compat/polling) the `polling` crate from the `smol-rs` ecosystem instead.
> 
> [tokio-rs/mio#1602](https://github.com/tokio-rs/mio/pull/1602) - see here. TL;DR: I'm still working on a fully proper implementation (though CI is the main problem right now), but I already have a fully working mio for ESP32 implementation (and thus Tokio for ESP32)

Fingers crossed for a successful merge upstream! Which reminds me that I should start the effort to upstream my own changes to `socket2` and `polling`. Which are - fortunately - very small, so hopefully will be accepted upstream.

**Comment by zvolin** [comment]
Hi,
I also came across this when trying to use `native-tls` enabled crates in [Intel SGX enclaves](https://github.com/fortanix/rust-sgx) environment. I decided to give this issue a try from where @MabezDev finished. Big thanks for this kickstart, as I'm not really familiar with TLS stuff etc.
I'm not sure if there is still an interest in this, but if so, I think I'll be able to issue some initial PR in a few days hopefully :crossed_fingers:

**Comment by cs-clarence** [comment]
really need this right now

**Comment by zvolin** [comment]
iirc it was working on the branch from my PR and the master of https://github.com/tiny-http/tiny-http (they didn't release a version with rust-native-tls support so far). Unfortunately there seems to be no interest in this change here

**Comment by ivmarkov** [comment]
I do realize it might not be a solution within the context of folks willing to reuse app-level Rust crates which are already hard-wired to `rust-native-tls`, but in the meantime `esp-idf-svc` has the `EspTls` blocking client TLS wrapper, as well as its `EspAsyncTls` `async` sibling. These are currently client TLS connection only, but a server one might be coming too. Which might happen faster if someone is willing to contribute a PR to `esp-idf-svc`. :)
